pageserver/src/tenant.rs

```diff
@@ -1350,13 +1267,72 @@ impl Tenant {
+                    // not-yet-uploaded one. it should be deleted now, because the branching might
+                    // not have been valid as it's ancestor may have been restored to earlier state
+                    // as well. in practice, control plane will keep retrying.
+                    //
+                    // first ensure that the un-uploaded timeline looks like it should, as in we
+                    // are not accidentially deleting a timeline which was ever active:
+                    // - root timelines have metadata and one possibly partial layer
+                    // - branched timelines have metadata
+                    //
+                    // if the timeline does not look like expected, fail loading of the tenant.
+                    // cleaning the timeline up manually and reloading the tenant is possible via
+                    // the above log message.
+                    let path = self.conf.timeline_path(&self.tenant_id, &timeline_id);
+
+                    let span = tracing::Span::current();
+
+                    return tokio::task::spawn_blocking({
+                        move || {
+                            use std::str::FromStr;
+                            use crate::tenant::storage_layer::LayerFileName;
+
+                            let _e = span.entered();
+                            let mut metadata = false;
+                            let mut layers = 0;
+                            let mut others = 0;
+                            for dentry in path.read_dir_utf8()? {
+                                let dentry = dentry?;
+                                let file_name = dentry.file_name();
+
+                                if file_name == "metadata" {
+                                    metadata = true;
+                                    continue;
+                                }
+
+                                if LayerFileName::from_str(file_name).is_ok()
+                                {
+                                    layers += 1;
+                                    continue;
+                                }
+
+                                others += 1;
+                            }
+
+                            // bootstrapped have the one image layer file, or one partial temp
+                            // file, branched have just the metadata
+                            if !(metadata && layers + others <= 1) {
+                                anyhow::bail!("unexpected assumed unuploaded, never been active timeline: found metadata={}, layers={}, others={}", metadata, layers, others);
+                            }
+
+                            let tmp_path =
+                                path.with_file_name(format!("{timeline_id}{}", TEMP_FILE_SUFFIX));
+                            std::fs::rename(path, &tmp_path)?;
+                            std::fs::remove_dir_all(&tmp_path)?;
```

Comment:
1. Missing an fsync of the parent directory after the rename.

2. What if we crash after the rename. We now have an non-TimelineId-conformant directory in the `timelines` dir. I guess the `scan_and_sort_timelines_dir` will treat it as `is_temporary` and delete? => worth a comment.

Also, we should just move the `safe_remove_tenant_dir_all` function from mgr.rs into `crashsafe.rs` and use it here as well (obviously need a sync version of it.)

Reply:
I don't think any fsyncs are needed here -- if the dir is present on next startup it'll just get removed again.

Reply:
Ok, I see it now.